sklearn/linear_model/_logistic.py

```diff
@@ -1586,6 +1207,10 @@ def fit(self, X, y, sample_weight=None):
             prefer = "threads"
         else:
             prefer = "processes"
+        if solver in ["lbfgs", "newton-cg"] and len(classes_) == 1:
+            n_threads = _openmp_effective_n_threads()
```

Comment:
Maybe, this should also depend on `n_jobs`, e.g. `n_threads = min(n_jobs, _openmp_effective_n_threads())`.

Reply:
Profiling on MNIST shaped problem shows that this does not seem a problem in practice. Maybe this is because the 2 OpenBLAS calls dominate the Cython calls that are comparatively negligible.

I only tried on macOS though. Maybe the OpenBLAS/OpenMP story is different on Linux.

Reply:
What is the conclusion? No code change needed?

Reply:
> No code change needed?

I don't think code change is required now to change the way Cython/prange/OpenMP threading is used in this PR but it might be important to keep this potential problem in mind if we observe slowe than expected performance on data with different shapes (e.g. smaller data size maybe where the interacting threadpool problem might still be present or in other locations in the scikit-learn code base whenever with have the following code pattern:

```
while not converged:
     step 1: Cython prange loop that uses an OpenMP threads
     step 2: numpy matrix multiply / scipy.linalg BLAS open that uses OpenBLAS threads
```

The long term fix for this kind of problems would ideally happen at the community level in upstream packages (https://github.com/scipy/scipy/issues/15129) but if ever we need to workaround it for a specific  in scikit-learn before that happens, we can always decide to use threadpoolctl to disable threading either for step 1 or step 2 when we detect that OpenBLAS is packaged in a way that it relies on its native threadpool instead of sharing the OpenMP threadpool.

Reply:
https://github.com/xianyi/OpenBLAS/issues/3187 relates to the issue @ogrisel mentions regarding the following code pattern.

Reply:
Actually, looking at this, and independently of the OpenBLAS / OpenMP threadpool interaction hypothetical problem, there might be problematic case when `n_jobs > 1` because we could get over-subscription between joblib parallelism and openmp parallelism. Maybe we should always hardcode `n_threads=1` to stay on the same side for this PR.

We might want to add a TODO comment to say that we should refactor this all to avoid using joblib parallelism entirely when doing binary and multinomial multiclass classification and only use joblib only for for the one-vs-rest multiclass case. This is probably beyond the scope of this PR.

Reply:
Done in https://github.com/scikit-learn/scikit-learn/pull/21808/commits/dea9bf099517ff85e767fdcc893fc8d24f6e521f (sorry for the stupid commit message).